modules/core/src/math/random-generator/vpUniRand.cpp

```diff
@@ -29,69 +29,186 @@
+  * You may obtain a copy of the License at
+  *
+  *     http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  *
+  * For additional information about the PCG random number generation scheme,
+  * including its license and other licensing options, visit
+  *
+  *     http://www.pcg-random.org
+  */
+
+  /*
+   * This code is derived from the full C implementation, which is in turn
+   * derived from the canonical C++ PCG implementation. The C++ version
+   * has many additional features and is preferable if you can use C++ in
+   * your project.
+   */
 
-#include <math.h>
 #include <visp3/core/vpUniRand.h>
 
+vpUniRand::vpUniRand() :
+  m_maxInvDbl(1.0 / UINT32_MAX), m_maxInvFlt(1.0f / UINT32_MAX), m_rng()
+{
+}
+
 /*!
-  Minimal random number generator of Park and Miller \cite Park:1988. Returns
-  a uniform random deviate between 0.0 and 1.0.
+  Create a pseudorandom number generator with uniform distribution.
+  \param seed : Starting state for the RNG, you can pass any 64-bit value.
+  \param seq : Select the output sequence for the RNG, you can pass any 64-bit value,
+  although only the low 63 bits are significant.
 
+  \sa setSeed
 */
-inline void vpUniRand::draw0()
+vpUniRand::vpUniRand(uint64_t seed, uint64_t seq) :
+  m_maxInvDbl(1.0 / UINT32_MAX), m_maxInvFlt(1.0f / UINT32_MAX), m_rng()
 {
-  long k = x / q; // temp value for computing without overflow
-  x = a * (x - k * q) - k * r;
-  if (x < 0)
-    x += m; // compute x without overflow
+  setSeed(seed, seq);
 }
 
 /*!
-  Bays-Durham Shuffling of Park-Miller generator.
+  Generates a pseudorandom uniformly distributed double number between [0, 1) range.
+  This is equivalent to call uniform(0.0, 1.0);
+*/
+double vpUniRand::operator()()
+{
+  return uniform(0.0, 1.0);
+}
 
-  Minimal random number generator of Park and Miller with Bays-Durham
-  shuffle. Returns a uniform random deviate between 0.0 and 1.0 (exclusive of
-  the endpoint values).
+/*!
+  Generates a uniformly distributed 32-bit unsigned integer less than bound
+  (i.e., x where 0 <= x < bound).
+
+  \note
+  <quote>
+  Some programmers may think that they can just run rng.next() % bound,
+  but doing so introduces nonuniformity when bound is not a power of two.
+  The code for boundedRand() avoids the nonuniformity by dropping a portion
+  of the RNG's output.
+  </quote>
 */
-double vpUniRand::draw1()
+uint32_t vpUniRand::boundedRand(uint32_t bound)
 {
-  const long ntab = 33; // we work on a 32 elements array.
-                        // the 33rd one is actually the first value of y.
-  const long modulo = ntab - 2;
-
-  static long y = 0;
-  static long T[ntab];
```

Comment:
These static variables store a global state that prevent to have repeatable results.